core/src/services/fs/writer.rs

```diff
@@ -101,3 +108,128 @@ impl oio::BlockingWrite for FsWriter<std::fs::File> {
         Ok(())
     }
 }
+
+impl oio::PositionWrite for FsWriter<tokio::fs::File> {
+    #[cfg(not(target_os = "windows"))]
+    async fn write_all_at(&self, offset: u64, buf: Buffer) -> Result<()> {
+        let f = self.f.as_ref().expect("FsWriter must be initialized");
+
+        let f = f
+            .try_clone()
+            .await
+            .map_err(new_std_io_error)?
+            .into_std()
+            .await;
+
+        tokio::task::spawn_blocking(move || {
+            f.write_all_at(&buf.to_bytes(), offset)
+                .map_err(new_std_io_error)
+        })
+        .await
+        .map_err(new_task_join_error)?
+    }
+
+    #[cfg(target_os = "windows")]
+    async fn write_all_at(&self, offset: u64, buf: Buffer) -> Result<()> {
+        let f = self.f.as_ref().expect("FsWriter must be initialized");
+
+        let f = f
+            .try_clone()
+            .await
+            .map_err(new_std_io_error)?
+            .into_std()
+            .await;
+
+        tokio::task::spawn_blocking(move || {
+            let mut offset = offset;
+            let mut buf: &[u8] = &buf.to_bytes();
+            while !buf.is_empty() {
+                match f.seek_write(buf, offset) {
+                    Ok(n) => {
+                        buf = &buf[n..];
+                        offset += n as u64
+                    }
+                    Err(e) => return Err(e).map_err(new_std_io_error),
+                }
+            }
+            Ok(())
+        })
+        .await
+        .map_err(new_task_join_error)?
+    }
+
+    #[cfg(target_os = "windows")]
+    async fn close(&self, offset: u64, buf: Buffer) -> Result<()> {
+        let f = self.f.as_ref().expect("FsWriter must be initialized");
+
+        let f = f
+            .try_clone()
+            .await
+            .map_err(new_std_io_error)?
+            .into_std()
+            .await;
+
+        tokio::task::spawn_blocking(move || {
+            let mut offset = offset;
+            let mut buf: &[u8] = &buf.to_bytes();
```

Comment:
`to_bytes` will copy the entire buffer to bytes which is slow. Maybe we can use:

```rust
while !buf.is_empty() {
    match f.seek_write(buf.chunk(), offset) {
        Ok(n) => {
            buf.advance(n);
            offset += n as u64
        }
        Err(e) => return Err(e).map_err(new_std_io_error),
    }
}
```